### PR TITLE
Kokkos promotion deprecation changes

### PR DESCRIPTION
**Pull request by crtrott**
Don't merge this in yet. We go test failures

```
76% tests passed, 46 tests failed out of 194

Label Time Summary:
Tpetra    = 1076.23 sec*proc (194 tests)

Total Test time (real) = 334.75 sec

The following tests FAILED:
	 11 - TpetraCore_gemm_m_eq_1_MPI_1 (Failed)
	 12 - TpetraCore_gemm_m_eq_2_MPI_1 (Failed)
	 13 - TpetraCore_gemm_m_eq_5_MPI_1 (Failed)
	 14 - TpetraCore_gemm_m_eq_13_MPI_1 (Failed)
	 15 - TpetraCore_BlockMultiVector_MPI_4 (Failed)
	 17 - TpetraCore_BlockCrsMatrix_MPI_4 (Failed)
	 19 - TpetraCore_BlockView_MPI_1 (Failed)
	 23 - TpetraCore_BlankRowBugTest_MPI_2 (Failed)
	 31 - TpetraCore_Core_initialize_where_user_initializes_mpi_MPI_4 (Failed)
	 32 - TpetraCore_Core_ScopeGuard_where_user_initializes_mpi_MPI_4 (Failed)
	 35 - TpetraCore_Core_initialize_where_tpetra_initializes_kokkos_MPI_1 (Failed)
	 36 - TpetraCore_Core_ScopeGuard_where_tpetra_initializes_kokkos_MPI_1 (Failed)
	 37 - TpetraCore_Core_initialize_where_user_initializes_kokkos_MPI_1 (Failed)
	 38 - TpetraCore_Core_ScopeGuard_where_user_initializes_kokkos_MPI_1 (Failed)
	 39 - TpetraCore_Core_initialize_where_tpetra_initializes_mpi_and_user_initializes_kokkos_MPI_2 (Failed)
	 40 - TpetraCore_Core_ScopeGuard_where_tpetra_initializes_mpi_and_user_initializes_kokkos_MPI_2 (Failed)
	 48 - TpetraCore_Issue601_MPI_4 (Failed)
	 53 - TpetraCore_CrsMatrix_UnitTests_MPI_4 (Failed)
	 54 - TpetraCore_CrsMatrix_UnitTests2_MPI_4 (Failed)
	 56 - TpetraCore_CrsMatrix_UnitTests4_MPI_4 (Failed)
	 57 - TpetraCore_CrsMatrix_NonlocalAfterResume_MPI_4 (Failed)
	 61 - TpetraCore_CrsMatrix_WithGraph_Serial_MPI_4 (Failed)
	 62 - TpetraCore_CrsMatrix_ReplaceDomainMapAndImporter_MPI_4 (Failed)
	 64 - TpetraCore_CrsMatrix_NonlocalSumInto_Ignore_MPI_4 (Failed)
	 70 - TpetraCore_CrsMatrix_ReplaceLocalValues_MPI_4 (Failed)
	 78 - TpetraCore_Albany182_MPI_4 (Failed)
	 83 - TpetraCore_Issue1752_MPI_2 (Failed)
	 95 - TpetraCore_ExportToStaticGraphCrsMatrix_MPI_4 (Failed)
	101 - TpetraCore_ImportExport2_UnitTests_MPI_4 (Failed)
	103 - TpetraCore_MatrixMarket_Tpetra_CrsMatrix_InOutTest_MPI_4 (Failed)
	104 - TpetraCore_MatrixMarket_Tpetra_CrsGraph_InOutTest_MPI_4 (Failed)
	105 - TpetraCore_MatrixMarket_Operator_Test_MPI_4 (Failed)
	112 - TpetraCore_CooMatrix_MPI_2 (Failed)
	136 - TpetraCore_MatrixMatrix_UnitTests_MPI_4 (Failed)
	137 - TpetraCore_AddProfiling_UnitTests_MPI_4 (Failed)
	139 - TpetraCore_MultiVector_UnitTests_MPI_4 (Failed)
	144 - TpetraCore_MultiVector_get2dView_MPI_1 (Failed)
	162 - TpetraCore_RowMatrixTransposer_test_MPI_4 (Failed)
	163 - TpetraCore_RowMatrixTransposer_UnitTests_MPI_4 (Failed)
	173 - TpetraCore_lesson05_redistribution_MPI_4 (Failed)
	176 - TpetraCore_FEMAssembly_InsertGlobalIndicesDP_MPI_4 (Failed)
	177 - TpetraCore_FEMAssembly_LocalElementLoopDP_MPI_4 (Failed)
	178 - TpetraCore_FEMAssembly_TotalElementLoopDP_MPI_4 (Failed)
	179 - TpetraCore_FEMAssembly_TotalElementLoopSP_MPI_4 (Failed)
	180 - TpetraCore_AdditiveSchwarzHalo_MPI_4 (Failed)
	194 - TpetraCore_guide_data_redist_1_MPI_4 (Failed)
```

### CODE REVIEW

**Comment by crtrott** [comment]
Ok the first problem is this:
```
 p=0: *** Caught standard std::exception of type 'std::runtime_error' :
 
  Deprecation Error: Kokkos::deep_copy extents of views don't match: C2(13,1) A_orig(1,13) 
  Traceback functionality not available
  
 [FAILED]  (0.213 sec) Blas_longlong_Gemm_UnitTest
 Location: /ascldap/users/crtrott/Trilinos/packages/tpetra/core/test/Blas/gemm.cpp:392
```
in the gemm tests. This actually never did what you thought it did. It would have only deep copied the (1,1) element. For Host builds it would have copied everything since both views are accessible from the host execution space, and thus a transpose deep copy can be done. But deep_copy used to simply take the minimum extents in cases where a bitwise copy was used. We deprecated that behavior ...

**Comment by ndellingwood** [comment]
@crtrott I have a couple more sacado fixes with deprecated code disabled that I'm going to add to this PR.

**Comment by ndellingwood** [comment]
Fixes pushed for `TpetraCore_BlockView_MPI_1`

**Comment by ndellingwood** [comment]
I tested `TpetraCore_BlockCrsMatrix_MPI_4` with promotion testing as well as in the Trilinos master branch and saw errors like this:

```
bash-4.2$ mpiexec -np 4 -bind-to socket -map-by socket ./TpetraCore_BlockCrsMatrix.exe

ibv_exp_query_device: invalid comp_mask !!! (comp_mask = 0x1003f84bdd0 valid_mask = 0x1)
[white22][[26293,1],1][btl_openib_component.c:1648:init_one_device] error obtaining device attributes for mlx5_0 errno says Invalid argument
--------------------------------------------------------------------------
WARNING: There was an error initializing an OpenFabrics device.

  Local host:   white22
  Local device: mlx5_0

```

Since this is on the Kepler cards, maybe there is some assumptions about Cuda-aware MPI that need to be disabled? Going to look through some recent issues about this and may reconfigure and test without this feature.

**Comment by srajama1** [comment]
@kyungjoo-kim faced this same error w/ Ifpack2 tests at some point. He might know a fix.

**Comment by ndellingwood** [comment]
Thanks @srajama1, will check in with him after another sanity check. Reconfiguring Trilinos master build with `Tpetra_ASSUME_CUDA_AWARE_MPI=OFF` and going to retest as sanity check. Cross-referencing #3405 in case it is relevant.

**Comment by ndellingwood** [comment]
Buliding this PR (and matching kokkos + kokkos-kernels branches) with `Tpetra_ASSUME_CUDA_AWARE_MPI=OFF` resulted in the test above running but segfaulting, so that's progress I spose...

Command and output:
```

bash-4.2$ mpiexec -np 4 -bind-to socket -map-by socket ./TpetraCore_BlockCrsMatrix.exe 
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white23 and rank 0!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white23 and rank 1!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white23 and rank 2!
Teuchos::GlobalMPISession::GlobalMPISession(): started processor with name white23 and rank 3!

***
*** Unit test suite ...
***

Sorting tests by group name then by the order they were added ... (time = 2.48e-05)

Running unit tests ...

0. BlockCrsMatrix_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_ctor_UnitTest ... [Passed] (0.719 sec)
1. BlockCrsMatrix_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_basic_UnitTest ... [Passed] (0.0947 sec)
2. BlockCrsMatrix_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_write_UnitTest ... --------------------------------------------------------------------------
mpiexec noticed that process rank 0 with PID 0 on node white23 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```

**Comment by ndellingwood** [comment]
`TpetraCore_BlockCrsMatrix.exe` (i.e. `TpetraCore_BlockCrsMatrix_MPI_4`) is dying first in the `write` test `BlockCrsMatrix_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_write_UnitTest`:

at this line:
`Tpetra::Experimental::blockCrsMatrixWriter(blockMat, "savedBlockMatrix.m", pl);`

**Comment by kyungjoo-kim** [comment]
@mhoemmen is blockcrsmatrix writer designed for working on gpus ? a natural logic seems to be mirror to host and write it to a stream (or file pointer). is this related to dualview changes ? I did not  check the code but the error type and the function name just lead to that suspect.

**Comment by ndellingwood** [comment]
Some fixes to the PR kokkos/kokkos#1860 fixed a lot of the failing tests posted [here](https://github.com/trilinos/Trilinos/pull/3682#issuecomment-431547502). My local build may be hanging on the Albany182 test, but all others up to that point are now passing.

**Comment by ndellingwood** [comment]
Problematic tests:

**`TpetraCore_Albany182_MPI_4`**
`CrsMatrix_int_longlong_default_node_type_Albany182_UnitTest`

**`TpetraCore_ImportExport2_UnitTests_MPI_4`**
`CrsMatrixImportExport_int_int_double_doImport_UnitTest` - seems to hang

**`TpetraCore_MatrixMatrix_UnitTests_MPI_4`**
`Tpetra_MatMat_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_operations_test_UnitTest`
`Tpetra_MatMat_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_operations_test_UnitTest`

**`TpetraCore_MultiVector_UnitTests_MPI_4`**
`MultiVector_int_int_double_Kokkos_Compat_KokkosSerialWrapperNode_ViewCtor_UnitTest MultiVector_int_longlong_double_Kokkos_Compat_KokkosSerialWrapperNode_ViewCtor_UnitTest `

**`TpetraCore_MultiVector_get2dView_MPI_1`**
`MultiVector_double_int_int_Kokkos_Compat_KokkosCudaWrapperNode_get2dView_noncontig_UnitTest MultiVector_double_int_longlong_Kokkos_Compat_KokkosCudaWrapperNode_get2dView_noncontig_UnitTest`

**`TpetraCore_RowMatrixTransposer_test_MPI_4`**

**`TpetraCore_FEMAssembly_*`**
These tests error out due to view ctors given more args than the rank

**`TpetraCore_AdditiveSchwarzHalo_MPI_4`**

In many cases errors related to use of `need_sync_host` emitted messages such as `The input Kokkos::DualView was most recently modified on device, but this function needs the host view of the data to be the most recently modified.` occurred.

**Comment by ndellingwood** [comment]
Down to 8 tests needing fixed, the FEM test should be trivial.

**Comment by ndellingwood** [comment]
Fixes pushed for `TpetraCore_FEMAssembly_*` tests.

**Comment by ndellingwood** [comment]
Test `TpetraCore_FEMAssembly_LocalElementLoopDP_MPI_4` still failing due to several errors of type: 
```
2: terminate called after throwing an instance of 'std::logic_error'
2:   what():  ../../packages/tpetra/core/src/Tpetra_Util.hpp:886:
2: 
2: Throw number = 1
2: 
2: Throw test that evaluated to true: x.need_sync_host ()
2: 
2: The input Kokkos::DualView was most recently modified on device, but this function needs the host view of the data to be the most recently modified.
```

**Comment by ndellingwood** [comment]
Various tests in `TpetraCore_ImportExport2_UnitTests_MPI_4` are hanging in the calls to `importAndFillCompleteCrsMatrix`, will see what I can find.

**Comment by ndellingwood** [comment]
I pushed another fix to kokkos/kokkos#1860. This fixed more of the tests, the following are still failing:

```
The following tests FAILED:
	 96 - TpetraCore_ImportExport2_UnitTests_MPI_4 (Failed)
	131 - TpetraCore_MatrixMatrix_UnitTests_MPI_4 (Failed)
	134 - TpetraCore_MultiVector_UnitTests_MPI_4 (Failed)
```

No more hanging tests. 3 more to go!

**Comment by ndellingwood** [comment]
Failure messages in tests above:

TpetraCore_ImportExport2_UnitTests_MPI_4
Specific failing tests and message:
FusedImportExport,MueLuStyle
`Constructor for Kokkos View '' has mismatched number of arguments. Number of arguments = 0 but dynamic rank = 1`

TpetraCore_MatrixMatrix_UnitTests_MPI_4
Specific failing tests and message:
`Constructor for Kokkos View '' has mismatched number of arguments. Number of arguments = 0 but dynamic rank = 1`

TpetraCore_MultiVector_UnitTests_MPI_4
Specific failing tests and message:
MultiVector_int_int_double_Kokkos_Compat_KokkosSerialWrapperNode_ViewCtor_UnitTest
```
  gblSuccess = 0 == 1 : FAILED ==> ../../packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3968
  We modified X_gbl in host memory, and sync'd to device memory, but X_lcl did not change!
  gblSuccess = 0 == 1 : FAILED ==> ../../packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3978
```

MultiVector_int_longlong_double_Kokkos_Compat_KokkosSerialWrapperNode_ViewCtor_UnitTest
```
  gblSuccess = 0 == 1 : FAILED ==> ../../packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3968
  We modified X_gbl in host memory, and sync'd to device memory, but X_lcl did not change!
  gblSuccess = 0 == 1 : FAILED ==> ../../packages/tpetra/core/test/MultiVector/MultiVector_UnitTests.cpp:3978
```

**Comment by ndellingwood** [comment]
The `TpetraCore_ImportExport2_UnitTests_MPI_4` is failing due to the `MatrixMatrix::Multiply` call, as are tests in `TpetraCore_MatrixMatrix_UnitTests_MPI_4`.
This error occurs in kokkos-kernels:
`kokkos-kernels/src/sparse/impl/KokkosSparse_spgemm_impl_def.hpp` line 183 - several views are being assigned views constructed with an empty label for the string and no extent arguments. Not sure why, but I'll add the fix to kokkos/kokkos-kernels#334

**Comment by ndellingwood** [comment]
After kokkos-kernels fixes, last failing test with disabled deprecated code is `TpetraCore_MultiVector_UnitTests_MPI_4`.

**Comment by ndellingwood** [comment]
Just pushed fix for the final broken test.

**Comment by ndellingwood** [comment]
@crtrott I'm going to remove the WIP label and merge into the kokkos-promotion branch.

File: packages/tpetra/core/src/Tpetra_Details_getDiagCopyWithoutOffsets_def.hpp [hunk 1]
```diff
@@ -130,8 +130,8 @@ class GetLocalDiagCopyWithoutOffsetsNotFillCompleteFunctor {
 
     // Side effects start below this point.
 
-    diag.template modify<Kokkos::HostSpace> ();
-    D_lcl_ = diag.template getLocalView<Kokkos::HostSpace> ();
+    diag.modify_host ();
+    D_lcl_ = diag.getDualView().view_host ();
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_Details_getDiagCopyWithoutOffsets_def.hpp]: Why call `getDualView()` here?  Why not just call `getLocalViewHost()`?  We should deprecate and remove getDualView() from MultiVector, since it exposes an implementation detail of how Tpetra implements dual view semantics.

**Comment by ndellingwood** [comment]
Merged into kokkos-promotion branch and updated kokkos-promotion with a merge of the develop branch. Closing this branch since changes are incorporated into the kokkos-promotion branch.

**Comment by ndellingwood** [comment]
Oop, guess I don't need to close since it's merged.

File: packages/tpetra/core/src/Tpetra_MultiVector_def.hpp [hunk 2]
```diff
@@ -908,8 +908,8 @@ namespace Tpetra {
     if (numSameIDs > 0) {
       const std::pair<size_t, size_t> rows (0, numSameIDs);
       if (copyOnHost) {
-        auto tgt_h = this->template getLocalView<HMS> ();
-        auto src_h = create_const_view (sourceMV.template getLocalView<HMS> ());
+        auto tgt_h = getDualView().view_host();
```
Comment (mhoemmen) [inline packages/tpetra/core/src/Tpetra_MultiVector_def.hpp]: It would be better to call `this->view_host()` directly.  In general, if we have the method on MultiVector, we should use the MultiVector method.  `Kokkos::DualView` is an implementation method, not the primary interface.

File: sampleScripts/Sandia-SEMS/test_tpetra_sandia [hunk 3]
```diff
@@ -277,13 +277,19 @@ elif [ "$MACHINE" = "white" ]; then
...
   #KDD 9/18  module load devpack/20180521/openmpi/3.1.0/gcc/7.2.0/cuda/9.2.88
   #KDD 9/18  This configuration seems to work
-  module load git/2.10.1
-  module load openmpi/2.1.2/gcc/7.2.0/cuda/9.2.88
-  module load cmake/3.9.6
-  module load openblas/0.2.20/gcc/7.2.0
-  module load boost/1.65.1/gcc/7.2.0
-  module load cuda/9.2.88
-  module load netcdf-exo/4.4.1.1/openmpi/2.1.2/gcc/7.2.0/cuda/9.0.176
+  #module load git/2.10.1
```
Comment (mhoemmen) [inline sampleScripts/Sandia-SEMS/test_tpetra_sandia]: We might want to separate these SEMS script changes into their own pull request.

**Comment by mhoemmen** [comment]
Uh oh, I think I missed the merge.  Ah well :-)

**Comment by ndellingwood** [comment]
@mhoemmen don't worry, it was a PR against the kokkos-promotion branch, not develop, so still time to work out some of these details :)

File: packages/tpetra/core/src/Tpetra_Details_getDiagCopyWithoutOffsets_def.hpp (hunk 1 above)
Comment (ndellingwood) [reply packages/tpetra/core/src/Tpetra_Details_getDiagCopyWithoutOffsets_def.hpp]: @crtrott I think you made the change Mark is asking about here, any reason to not use `getLocalViewHost`?

**Comment by mhoemmen** [comment]
Thanks @ndellingwood ! :D

File: packages/tpetra/core/src/Tpetra_Details_getDiagCopyWithoutOffsets_def.hpp (hunk 1 above)
Comment (ndellingwood) [reply packages/tpetra/core/src/Tpetra_Details_getDiagCopyWithoutOffsets_def.hpp]: @mhoemmen fixed this case in the kokkos-promotion branch, there may be a few more to look out for on that branch.
